oak/server/oak_runtime.cc

```diff
@@ -94,40 +97,34 @@ void OakRuntime::SetUpChannels() {
   });
   // TODO: join() instead when we have node termination
   t.detach();
+}
 
-  // Create the channels needed for gRPC interactions.
-
+// Create the channels needed for gRPC interactions.
+void OakRuntime::SetUpGrpcChannels(OakNode& node) {
   // Incoming request channel: keep the write half in |OakRuntime|, but map
   // the read half to a well-known channel handle on |node_|.
   grpc_in_ = std::make_shared<MessageChannel>();
-  node_->SetChannel(ChannelHandle::GRPC_IN, absl::make_unique<MessageChannelReadHalf>(grpc_in_));
+  node.SetChannel(ChannelHandle::GRPC_IN, absl::make_unique<MessageChannelReadHalf>(grpc_in_));
   LOG(INFO) << "Created gRPC input channel: " << ChannelHandle::GRPC_IN;
 
   // Outgoing response channel: keep the read half in |OakRuntime|, but map
   // the write half to a well-known channel handle on |node_|.
   grpc_out_ = std::make_shared<MessageChannel>();
-  node_->SetChannel(ChannelHandle::GRPC_OUT, absl::make_unique<MessageChannelWriteHalf>(grpc_out_));
+  node.SetChannel(ChannelHandle::GRPC_OUT, absl::make_unique<MessageChannelWriteHalf>(grpc_out_));
   LOG(INFO) << "Created gRPC output channel: " << ChannelHandle::GRPC_IN;
+
+  grpc_node_->SetupGrpcChannels(grpc_in_, grpc_out_);
 }
 
-void OakRuntime::CompletionQueueLoop() {
-  LOG(INFO) << "Starting gRPC completion queue loop";
-  // The stream object will delete itself when finished with the request,
-  // after creating a new stream object for the next request.
-  auto stream =
-      new ModuleInvocation(module_service_.get(), completion_queue_.get(), grpc_in_, grpc_out_);
-  stream->Start();
-  while (true) {
-    bool ok;
-    void* tag;
-    if (!completion_queue_->Next(&tag, &ok)) {
-      LOG(FATAL) << "Failure reading from completion queue";
-      return;
-    }
-    auto callback = static_cast<std::function<void(bool)>*>(tag);
-    (*callback)(ok);
-    delete callback;
+int32_t OakRuntime::GetPort() { return grpc_node_->GetPort(); }
+
+asylo::Status OakRuntime::Finalize() {
+  if (grpc_node_) {
```

Comment:
And also stop all the other nodes? In fact, perhaps rename the `Stop` method to `Shutdown` as well, and leave a TODO to implement it.

Reply:
Done.